src/batch/src/executor/join/lookup_join.rs

```diff
@@ -0,0 +1,592 @@
+    /// that eill be used for sending RPC calls. They are made using the `source_templates` in the
+    /// passed `ProbeSideSourceParams`.
+    fn new_boxed(cur_row: &RowRef, params: &ProbeSideSourceParams) -> Result<Box<Self>> {
+        let Plan(inner_template_plan) =
+            params.source_templates[0].get_local_execute_plan().unwrap();
+
+        let local_execute_plan = LocalExecutePlan {
+            plan: Some(PlanFragment {
+                root: Some(PlanNode {
+                    children: vec![],
+                    identity: Uuid::new_v4().to_string(),
+                    node_body: Some(DefaultGatherExecutor::create_row_seq_scan_node(
+                        cur_row, params,
+                    )?),
+                }),
+                exchange_info: None,
+            }),
+            epoch: inner_template_plan.epoch,
+        };
+
+        let prost_sources = params
+            .source_templates
+            .iter()
+            .map(|source| ProstExchangeSource {
+                task_output_id: Some(source.get_task_output_id().unwrap().clone()),
+                host: Some(source.get_host().unwrap().clone()),
+                local_execute_plan: Some(Plan(local_execute_plan.clone())),
+            })
+            .collect();
+
+        Ok(Box::new(DefaultGatherExecutor { prost_sources }))
+    }
+
+    /// Execute the `GatherExecutor` and retrieve the data chunk stream
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.get_probe_side_chunk()
+    }
+}
+
+/// Lookup Join Executor.
+/// High-level idea:
+/// 1) Iterate through each row in the build side
+/// 2) For each row R, find the rows that match with R on the join conditions on the probe side by
+/// using the `GatherExecutor` to query the probe side
+/// 3) Join R and the matching rows on the probe side together with either inner or left outer join
+/// 4) Repeat 2-3) for every row in the build side
+pub struct LookupJoinExecutor<GE> {
+    join_type: JoinType,
+    condition: Option<BoxedExpression>,
+    build_side_source: RowLevelIter,
+    build_side_data_types: Vec<DataType>,
+    probe_side_data_types: Vec<DataType>,
+    probe_side_source_params: ProbeSideSourceParams,
+    chunk_builder: DataChunkBuilder,
+    schema: Schema,
+    output_indices: Vec<usize>,
+    last_chunk: Option<SlicedDataChunk>,
+    identity: String,
+    unused: PhantomData<GE>, // Only exists so we can use generic type
+}
+
+impl<GE: 'static + GatherExecutor> Executor for LookupJoinExecutor<GE> {
+    fn schema(&self) -> &Schema {
+        &self.schema
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
+
+impl<GE: 'static + GatherExecutor> LookupJoinExecutor<GE> {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        self.build_side_source.load_data().await?;
```

Comment:
The reason NestedLoopJoin loads all data first because it need to join against probe data repeatedly. But in index lookup join we don't need this, so I prefer to join data row by row.